src/generic/DirectSum.jl

```diff
@@ -122,39 +137,31 @@ end
 #
 ###############################################################################
 
-function direct_sum_injection(m::AbstractAlgebra.FPModule{T}, D::DirectSumModule{T}, v::AbstractAlgebra.FPModuleElem{T}) where T <: RingElement
+function direct_sum_injection(i::Int, D::DirectSumModule{T}, v::AbstractAlgebra.FPModuleElem{T}) where T <: RingElement
+   S = summands(D)
+   m = S[i]
    R = base_ring(m)
    # Find starting point of the given module in the large vectors
-   start = 0
-   S = summands(D)
-   i = 1
-   while i < length(S) && S[i] !== m
-         start += ngens(S[i])
-         i += 1
-   end
+   start = sum(map(x->ngens(x)::Int, S[1:i-1]))
    # create embedded value
-   newv = T[zero(R) for i in 1:ngens(D)]
-   for i = 1:ngens(m)
-      newv[i + start] = v[i]
+   newv = T[zero(R) for j in 1:ngens(D)]
+   for j = 1:ngens(m)
+      newv[j + start] = v[j]
    end
    matv = matrix(R, 1, length(newv), newv)
    return DirectSumModuleElem{T}(D, matv)
 end
 
-function direct_sum_projection(D::DirectSumModule{T}, m::U, v::AbstractAlgebra.FPModuleElem{T}) where {T <: RingElement, U <: AbstractAlgebra.FPModule{T}}
+function direct_sum_projection(D::DirectSumModule{T}, i::Int, v::AbstractAlgebra.FPModuleElem{T}) where {T <: RingElement}
```

Comment:
Doc changes presumably needed.

Reply:
Also, do we need additional tests? E.g. a regression test would be useful, which fails on the old version.